dissect/volume/ddf/ddf.py

```diff
@@ -0,0 +1,389 @@
+
+    def __repr__(self) -> str:
+        return f"<VirtualDiskRecord guid={self.guid} number={self.number} type={self.type:#x} name={self.name!r}>"
+
+
+class VirtualDiskConfigurationRecord:
+    def __init__(self, fh: BinaryIO, num_entries: int):
+        self.header = c_ddf.VD_Configuration_Record(fh)
+
+        self.guid = self.header.VD_GUID
+        self.guid_timestamp = ts.from_unix(DECADE + int.from_bytes(self.guid[16:20], "little"))
+        self.timestamp = ts.from_unix(DECADE + self.header.Timestamp)
+        self.sequence_number = self.header.Sequence_Number
+        self.primary_element_count = self.header.Primary_Element_Count
+        self.strip_size = self.header.Strip_Size
+        self.primary_raid_level = self.header.Primary_RAID_Level
+        self.raid_level_qualifier = self.header.RAID_Level_Qualifier
+        self.secondary_element_count = self.header.Secondary_Element_Count
+        self.secondary_element_seq = self.header.Secondary_Element_Seq
+        self.secondary_raid_level = self.header.Secondary_RAID_Level
+        self.size = self.header.VD_Size
+        self.block_size = self.header.Block_Size
+        self.rotate_parity_count = self.header.Rotate_Parity_Count
+
+        self.physical_disk_sequence = c_ddf.uint32[num_entries](fh)[: self.primary_element_count]
+        self.starting_block = c_ddf.uint64[num_entries](fh)[: self.primary_element_count]
+
+    def __repr__(self) -> str:
+        return f"<VirtualDiskConfigurationRecord guid={self.guid}>"
+
+
+def _read_section_header(fh: BinaryIO, structure: Structure, signature: int) -> Instance:
+    obj = structure(fh)
+    if obj.Signature != signature:
+        raise DDFError(f"Invalid {structure.name} header. Signature: {obj.Signature:#010x}, expected {signature:#x}")
+    return obj
+
+
+def _read_physical_disk_records(fh: BinaryIO, offset: int) -> list[PhysicalDiskRecord]:
+    fh.seek(offset)
+    header = _read_section_header(fh, c_ddf.Physical_Disk_Records, c_ddf.DDF_PDR_SIGNATURE)
+    return [PhysicalDiskRecord(fh) for _ in range(header.Populated_PDEs)]
+
+
+def _read_virtual_disk_records(fh: BinaryIO, offset: int) -> list[VirtualDiskRecord]:
+    fh.seek(offset)
+    header = _read_section_header(fh, c_ddf.Virtual_Disk_Records, c_ddf.DDF_VD_RECORD_SIGNATURE)
+    return [VirtualDiskRecord(fh) for _ in range(header.Populated_VDEs)]
+
+
+def _read_virtual_disk_configuration_records(
+    fh: BinaryIO, offset: int, count: int, size: int, num_entries: int
+) -> list[VirtualDiskConfigurationRecord]:
+    result = []
+    for _ in range(count):
+        fh.seek(offset)
+        obj = VirtualDiskConfigurationRecord(fh, num_entries)
+
+        if obj.header.Signature != c_ddf.DDF_VDCR_SIGNATURE:
+            continue
+
+        result.append(obj)
+        offset += size
+
+    return result
+
+
+def _convert_raid_layout(prl: int, rlq: int, srl: int, pec: int, sec: int) -> tuple[Level, Layout, int]:
```

Comment:
Not sure how you'd split this up. Right now I like how it's just kind of separate.

Reply:
It is just a matter of style of course, but if you attach the method like I said (and add some @property decorators) you can write:

```
class DDFVirtualDisk(VirtualDisk):
    def __init__(
        self,
        vdcr: VirtualDiskConfigurationRecord,
        vdr: VirtualDiskRecord,
        disks: dict[int, tuple[int, DDFPhysicalDisk]],
    ):
        self.vdcr = vdcr
        self.vdr = vdr
        self.disks = disks

        if (block_size := self.vdcr.block_size) == 0xFFFF:
            block_size = list(self.disks.values())[0][1].block_size

        super().__init__(
            self.vdr.name,
            self.vdr.guid,
            self.vdcr.size * block_size,
            self.vdcr.level,
            self.vdcr.layout,
            (2**self.vdcr.strip_size) * block_size,
            self.vdcr.num_disks,
            disks,
        )
```

which I think is really elegant, because you have no 

`level, layout, num_disks = _convert_raid_layout(...) `

or do I miss something here?

Reply:
Ah I see. That does look a bit nicer, but my only qualm with that is that `VirtualDiskConfigurationRecord` is a "true DDF" class, and `DDFVirtualDisk` is the bridge between DDF and the generic `VirtualDisk` class. Putting the conversion in there feels more appropriate, personally, because that's where we're "converting" between DDF and the generic classes.

Reply:
I don't see a problem with this, as long as one object does not explicitly reference the other? But like I said, it's just a matter of style.